### PR TITLE
refactor: convert TableBodyWrapper to typescript

### PR DESCRIPTION
**Pull request by Donya-qlik**

### CODE REVIEW

File: src/table/components/TableBodyWrapper.tsx [hunk 1]
```diff
@@ -20,7 +20,7 @@ function TableBodyWrapper({
   tableWrapperRef,
   announce,
   children,
-}) {
+}: RootProps) {
```
Comment (LiKang6688) [inline src/table/components/TableBodyWrapper.tsx:23]: Create a BodyWrapperProps and extends RootProps

File: src/types.ts [hunk 2]
```diff
@@ -312,6 +312,9 @@ export interface RootProps {
   keyboard: stardust.Keyboard;
   footerContainer?: HTMLElement;
   announce: Announce;
+  setShouldRefocus(): void;
+  tableWrapperRef: React.MutableRefObject<HTMLDivElement | undefined>;
+  children: JSX.Element;
```
Comment (LiKang6688) [inline src/types.ts:317]:
```suggestion
export interface BodyWrapperProps extends Omit<RootProps, 'selectionsAPI'> {
  selectionsAPI: ExtendedSelectionAPI;
  setShouldRefocus(): void;
  tableWrapperRef: React.MutableRefObject<HTMLDivElement | undefined>;
  children: JSX.Element;
}
```

File: src/types.ts [hunk 3]
```diff
@@ -298,7 +298,7 @@ export interface RenderProps {
 
 export interface RootProps {
   direction?: 'ltr' | 'rtl';
-  selectionsAPI: stardust.ObjectSelections;
+  selectionsAPI: ExtendedSelectionAPI;
```
Comment (LiKang6688) [inline src/types.ts:301]: I am not sure when the stardust.ObjectSelections becomes the ExtendedSelectionAPI. Looking into it.

**Comment by LiKang6688** [comment]
I got the type of selectionsAPI wrong in the index.ts, should be:
`const selectionsAPI = useSelections() as ExtendedSelectionAPI;`

File: src/table/components/TableBodyWrapper.tsx [hunk 4]
```diff
@@ -31,15 +31,22 @@ function TableBodyWrapper({
...
     [columnsStylingIDsJSON, isSelectionsEnabled]
   );
   const bodyCellStyle = useMemo(() => getBodyCellStyle(layout, theme), [layout, theme]);
   const hoverEffect = layout.components?.[0]?.content?.hoverEffect;
   const cellStyle = { color: bodyCellStyle.color, backgroundColor: theme.table.backgroundColor };
-
   useEffect(() => {
-    addSelectionListeners({ api: selectionsAPI, selectionDispatch, setShouldRefocus, keyboard, tableWrapperRef });
+    addSelectionListeners({
+      api: selectionsAPI as ExtendedSelectionAPI,
```
Comment (Donya-qlik) [inline src/table/components/TableBodyWrapper.tsx:44]: @LiKang6688  set `api `as `selectionAPI ` is the reason of using ExtendedSelectionAPI here.

File: src/types.ts (hunk 3 above)
Comment (Donya-qlik) [reply src/types.ts:301]: I explained it at above on the right file.

**Comment by Donya-qlik** [comment]
> I got the type of selectionsAPI wrong in the index.ts, should be: `const selectionsAPI = useSelections() as ExtendedSelectionAPI;` all `stardust.ObjectSelections` should be replaced with `ExtendedSelectionAPI`

@LiKang6688 index.d.ts is the file that should be modified by the new type and as you know is not editable via the current PR, could you give me a hint how I could I fix those in nebula? there are couple of more things that should be change there that already has ts-ignore tag on.

**Comment by LiKang6688** [comment]
> > I got the type of selectionsAPI wrong in the index.ts, should be: `const selectionsAPI = useSelections() as ExtendedSelectionAPI;` all `stardust.ObjectSelections` should be replaced with `ExtendedSelectionAPI`
> 
> @LiKang6688 index.d.ts is the file that should be modified by the new type and as you know is not editable via the current PR, could you give me a hint how I could I fix those in nebula? there are couple of more things that should be change there that already has ts-ignore tag on.

@Donya-qlik You do not need to fix those in the nebula. 
We could do the fix in the src/index.ts(not index.d.ts), 
from
`const selectionsAPI = useSelections()`
to
`const selectionsAPI = useSelections() as ExtendedSelectionAPI;`

**Comment by Donya-qlik** [comment]
> as ExtendedSelectionAPI

Ah, true! Will fix it

**Comment by LiKang6688** [comment]
stardust.ObjectSelections should also be migrated to ExtendedSelectionAPI everywhere, for example, https://github.com/qlik-oss/sn-table/blob/main/src/types.ts#L276

File: src/types.ts [hunk 5]
```diff
@@ -315,9 +315,16 @@ export interface RootProps {
   announce: Announce;
 }
 
+export interface BodyWrapperProps
+  extends Omit<RootProps, 'changeSortOrder' | 'rect' | 'pageInfo' | 'setPageInfo' | 'footerContainer' | 'translator'> {
```
Comment (haxxmaxx) [inline src/types.ts:319]: Not sure about this.  Almost half the props on RootProps are being omitted and three are added. As I mentioned in here https://github.com/qlik-oss/sn-table/pull/663 I'm not sure we should omit these as they are actually passed.

File: src/table/components/__tests__/TableBodyWrapper.spec.tsx [hunk 6]
```diff
@@ -38,49 +55,58 @@ describe('<TableBodyWrapper />', () => {
...
+    tableFirstRow = tableData.rows[0]['col-0'] as Cell;
   });
 
   afterEach(() => jest.clearAllMocks());
 
   it('should render 2x2 table body and call CellRenderer', () => {
     jest.spyOn(getCellRenderer, 'default').mockImplementation(() => {
       cellRendererSpy();
-      return 'td';
+      return 'td' as unknown as CellHOC;
```
Comment (haxxmaxx) [inline src/table/components/__tests__/TableBodyWrapper.spec.tsx:92]: I think this is way more complicated than needed. Try just `jest.spyOn(getCellRenderer, 'default')` and then `    expect(getCellRenderer.default).toHaveBeenCalledTimes(2);`

Comment (haxxmaxx) [reply src/table/components/__tests__/TableBodyWrapper.spec.tsx:92]: Goes for the other tests below as well

**Comment by haxxmaxx** [comment]
Looks good but seems like you need to fix those rendering tests

File: src/types.ts (hunk 5 above)
Comment (Donya-qlik) [reply src/types.ts:319]: First I added them to RootProps but Li suggested the extended ones. The only benefit of omitting the unnecessary ones is we don't need to add every single RootProps in TableBodyWraper test component.

**Comment by Donya-qlik** [comment]
> Looks good but seems like you need to fix those rendering tests

@haxxmaxx rendering tests? There is no test failure as far as I see ...

File: src/table/components/__tests__/TableBodyWrapper.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    jest.spyOn(handleKeyPress, 'handleBodyKeyDown');
+
+    const { getByText } = renderTableBody();
+    fireEvent.keyDown(getByText(tableFirstRow.qText as string));
+
+    expect(handleKeyPress.handleBodyKeyDown).toHaveBeenCalledTimes(1);
+  });
+
+  it('should call handleClickToFocusBody on mouseDown', () => {
+    jest.spyOn(handleAccessibility, 'handleClickToFocusBody').mockImplementation(() => jest.fn());
```
Comment (haxxmaxx) [inline src/table/components/__tests__/TableBodyWrapper.spec.tsx:108]:
```suggestion
    jest.spyOn(handleAccessibility, 'handleClickToFocusBody');
```

File: src/table/components/__tests__/TableBodyWrapper.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+          tableWrapperRef={tableWrapperRef}
+          announce={announce}
+        >
+          {children}
+        </TableBodyWrapper>
+      </TableContextProvider>
+    );
+
+  beforeEach(async () => {
+    jest.spyOn(selectionsUtils, 'addSelectionListeners').mockImplementation(() => jest.fn());
```
Comment (haxxmaxx) [inline src/table/components/__tests__/TableBodyWrapper.spec.tsx:63]:
```suggestion
    jest.spyOn(selectionsUtils, 'addSelectionListeners');
```

File: src/types.ts (hunk 5 above)
Comment (haxxmaxx) [reply src/types.ts:319]: ok go for this now and we might refactor it later

**Comment by haxxmaxx** [comment]
> > Looks good but seems like you need to fix those rendering tests
> 
> @haxxmaxx rendering tests? There is no test failure as far as I see ...

I must have seen a ghost. Tests looks fine!

File: src/table/components/__tests__/TableBodyWrapper.spec.tsx (hunk 8 above)
Comment (Donya-qlik) [reply src/table/components/__tests__/TableBodyWrapper.spec.tsx:63]: There is a bunch of `api.on( ... )` lines in `addSelectionListeners` that by this change,  are not functioning, seems this line is needed.

Comment (haxxmaxx) [reply src/table/components/__tests__/TableBodyWrapper.spec.tsx:63]: that makes sense actually. Try with
```suggestion
    jest.spyOn(selectionsUtils, 'addSelectionListeners').mockImplementation();
```

**Review by haxxmaxx — APPROVED** [review]

**Pull request merged**